consensus/src/simplex/actors/voter/slot.rs

```diff
@@ -0,0 +1,359 @@
+        self.proposal = Some(proposal);
+        self.status = Status::Verified;
+        self.requested_build = true;
+        self.requested_verify = true;
+    }
+
+    pub fn request_verify(&mut self) -> bool {
+        if self.requested_verify {
+            return false;
+        }
+        self.requested_verify = true;
+        true
+    }
+
+    pub fn mark_verified(&mut self) -> bool {
+        if self.status != Status::Unverified {
+            return false;
+        }
+        self.status = Status::Verified;
+        true
+    }
+
+    /// Marks the slot as waiting on parent certificates.
+    ///
+    /// Returns `true` the first time it is invoked so callers can distinguish
+    /// between a freshly-discovered gap (which should trigger a resolver fetch)
+    /// and repeated checks we expect to run while we wait for the data.
+    pub fn mark_parent_missing(&mut self, parent: View) -> bool {
+        match self.awaiting_parent {
+            Some(missing) if missing == parent => false,
+            None | Some(_) => {
+                self.awaiting_parent = Some(parent);
+                true
+            }
+        }
+    }
+
+    pub fn clear_parent_missing(&mut self) {
+        self.awaiting_parent = None;
+    }
+
+    pub fn update(&mut self, proposal: &Proposal<D>, recovered: bool) -> Change<D> {
+        // Once we mark the slot as replaced we refuse to record any additional
+        // votes, even if they target the original payload. Unless there is
+        // a safety failure, we won't be able to use them for anything so we might
+        // as well ignore them.
+        if self.status == Status::Replaced {
+            return Change::Skipped;
+        }
+        match &self.proposal {
+            None => {
+                self.proposal = Some(proposal.clone());
+                self.status = if recovered {
+                    Status::Verified
+                } else {
+                    Status::Unverified
+                };
+                Change::New
+            }
+            Some(existing) if existing == proposal => {
+                if recovered {
+                    self.status = Status::Verified;
+                }
+                Change::Unchanged
+            }
+            Some(existing) => {
+                let mut dropped = existing.clone();
+                let mut retained = proposal.clone();
+                if recovered {
+                    // If we receive a certificate for a conflicting proposal, we replace the
+                    // local proposal (may just be from a vote)
+                    self.proposal = Some(retained.clone());
+                    self.requested_build = true;
+                    self.requested_verify = true;
+                } else {
+                    // If this isn't a certificate, we keep the proposal as-is
+                    retained = existing.clone();
+                    dropped = proposal.clone();
```

Comment:
nit: are you able to do something like `(retained, dropped) = (dropped, retained);` instead of cloning both again?